### PR TITLE
Adds a skill system

### PR DESCRIPTION
**Pull request by nmajask**
# Document the changes in your pull request

Adds a skill system which is heavily inspired by Baystation, various RGPs and some of the implementation by Hopek in #10802.
There are 23 skills, each of which will be mentioned in a list below, and each skill possesses 5 levels, with each consecutive level being more difficult or costly to obtain. 
Players will have a pool of points with which to purchase these skills, and differing jobs will have differing base proficiencies (ex. Scientist will have a better starting Design skill than an Assistant). These skills and levels will have varying effects on relevant actions, and/or can offer bonuses depending on the situation.

## New Skills
Obviously, this adds new skills, 23 to be exact. Most are useless right now, but I do plan on fixing that later.

- Strength: How good you are at lifting, pushing, or carrying heavy objects.
- Dexterity: How good you are at quickly and accurately moving yourself and your limbs.
- Endurance: How good you are at tanking through hits and resisting chemicals and alcohol.
- Botany: How good you are at planting and tending to plants.
- Cooking: How good you are using culinary tools.
- Creativity: How creative you can be.
- Survival: How well you can survive by yourself.
- Piloting: How good you are at piloting mechs or other vehicles.
- Leadership: How good you are at leading and inspiring others.
- Hand-to-Hand: How good you are at fighting with your bare hands.
- Melee Weapons: How good you are at fighting with a toolbox or batlong.
- Ranged Weapons: How good you are at fighting with a gun.
- Forensics: How good you are at finding evidence from the environment.
- Medicine: How good you are at identifying and treating injuries.
- Anatomy: How good you are with the innards of humans.
- Chemistry: How good you are at mixing and identifying chemicals.
- Psychology: How good you are at identifying the feelings of others.
- Design: How good you are at researching new designs.
- Information Technology: How good you are at using and creating computers.
- Biology: How knowledgeable you are at understanding how life works.
- Mechanics: How good you are at creating and manipulating machines.
- Electronics: How good you are at creating and manipulating electronics.
- Atmospherics: How good you are at identifying and manipulating gasses and atmospheric tools.

## Why?

- Makes power gaming dependent on code instead of the interpretation of a rule by admins
- Allows for more interesting characters
- Increases RP

## To-Do

- [ ] Make the skills useful
- [ ] Balance stuff
- [ ] Fix some of the UI
- [ ] Get better sprites for the items and skill button
- [ ] Clean up some of this code spaghetti 

## Current Uses For Skills

- [ ] Strength: None.
- [ ] Dexterity: None.
- [ ] Endurance: None.
- [ ] Botany: None.
- [ ] Cooking: Some crafting recipes.
- [ ] Creativity: None.
- [ ] Survival: Using some tools and in some crafting recipes.
- [ ] Piloting: None.
- [ ] Leadership: None.
- [ ] Hand-to-Hand: None.
- [ ] Melee Weapons: None.
- [ ] Ranged Weapons: Accuracy for most guns.
- [ ] Forensics: Using forensic scanner.
- [x] Medicine: Using medical scanner, identifying wounds, and some surgeries.
- [ ] Anatomy: Most surgeries.
- [ ] Chemistry: Increases the speed of transferring reagents.
- [ ] Psychology: None.
- [ ] Design: Some crafting recipes.
- [ ] Information Technology: None.
- [ ] Biology: None.
- [ ] Mechanics: Using some tools and in some crafting recipes.
- [ ] Electronics: Using some tools and in some crafting recipes.
- [ ] Atmospherics: None.

## Screenshots

image
image

# Wiki Documentation

Will probably need a whole new page for this, plus some information about skill proficiency's in the appropriate job page. 

# Changelog

:cl:  
rscadd: Added a new skill system
tweak: Scanning someone with a health analyzer requires a Basic medical skill and takes time if your skill level is bellow Experienced.
tweak: Transferring reagents now takes time depending on your chemistry level.
tweak: Wounds now require certain skills to be identified or give detains when inspected.
tweak: Surgeries now have an increased chance to fail if your level in the required skill is low.
tweak: Tools are now slower if your level in the required skill is low and faster if its high.
/:cl:

### CODE REVIEW

**Comment by ToasterBiome** [comment]
How is this different from the last time it was tried?

I personally think this is cool, but some people seemed to disagree last time

**Comment by adamsong** [comment]
Is there an admin skills panel? For bussing purposes

**Comment by nmajask** [comment]
> How is this different from the last time it was tried?
> 
> I personally think this is cool, but some people seemed to disagree last time

If you're talking about #10802, then the major difference is that this allows you to customize your skills instead of it being determined by your job

> Is there an admin skills panel? For bussing purposes

You can edit the skills directly in the skills menu, though I still need to allow admins to open it on other people.

**Comment by LazennG** [comment]
cool concept, hoping that the ability to choose skills doesn't mean that like 4 become meta and all others are useless, at least in the long run should this get merged

**Comment by maxion12345** [comment]
me min-maxing dexterity and melee for maximum shitcurity

Also please do into detail on what the actual skills will change in-game. Is strength related to fireman carry speed, or gives increased punch damage? Does survival grant you better resistance to wounds or something?

**Comment by bruhlookatthisdood** [comment]
Will there be quirks affecting the skills? Like decreased strength or something like that?

File: code/__DEFINES/{yogs_defines}/skills.dm [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+#define SKILLLEVEL_EXPERIENCED	4
+#define SKILLLEVEL_MASTER		5
+
+//Skill Difficulty Defines
+#define SKILLDIF_EASY			1
+#define SKILLDIF_MEDIUM			2
+#define SKILLDIF_HARD			3
+
+//Physical Skills
+#define SKILL_STRENGHT			"Strength"
```
Comment (adamsong) [inline code/__DEFINES/{yogs_defines}/skills.dm]: STRENGHT?

Comment (nmajask) [reply code/__DEFINES/{yogs_defines}/skills.dm]: Cant speel

**Comment by Marmio64** [comment]
either make the skills unable to benefit players in a combat sort of way or simply go with (what i personally prefer) an inherent skill system, like what we kind of already have, where certain jobs would get bonuses that go along with what they need to do. Problematic skills i see so far-
Dexterity: i don't know what this is implying, are you moving faster or what? How do you 'move your limbs faster', what, is it seriously going to buff attack speed? Either way, if its going to boost your movement or attack speed this is a straight power game skill.
Endurance: literally just a straight power game skill that makes you harder to kill
Hand-to-Hand: literally just straight power game skill
Melee Weapons: literally just straight power game skill
Ranged Weapons: literally just straight power game skill

Possibly problematic?-
Creativity & Survival: I have no idea what this implies but if survival is going to boost you in ways other than just like slowing hunger loss and if creativity will unlock weapons, that doesn't sound good to me.
Strength: probably would be abused by power gamers to make it faster to drag victims away.

The other skills of course can still be used in a way to benefit yourself, maybe even in a combat setting, but the ones i listed above seem way too combat and powergame focused and will likely be minmaxxed if left in, leaving the misc skills like 'Cooking' in the dust. Anyways, I don't want to be a heckler but there is a lot of problems with how this is looking so far and i've seen it in other servers, especially servers that use this skill system but aren't strictly HRP. For example, deadspace 13 is supposed to be decently high rp but after the halfway mark it turns into a TDM server where everyone minmaxes the martial arts skill to body slam necromorphs. Another solution would be to kind of take what i said earlier and to lock the skills from being chosen, but still keep in the skill system so that chefs could have better cooking skills, sec gets better fighting skills, etc, so you cant powergame to be another job or to be a combat god as some random scientist.


TL;DR : Too many powergame combat skills, seen it before in other non-HRP servers, its always minmaxed. Either completely remove the combat-focused skills or stop them from being player pickable and have jobs simply start with the skills they need.

**Comment by ghost** [comment]
This is a bad idea. It'll be abused for maximum powergaming ability by some, and those who need the points to do actual jobs will be at a disadvantage.

**Comment by ilikesaying** [comment]
> For example, deadspace 13 is supposed to be decently high rp but after the halfway mark it turns into a TDM server where everyone minmaxes the martial arts skill to body slam necromorphs.

Don't forget them also blocking half of the hits because rng blocking chance from high martial arts skills

**Comment by nmajask** [comment]
> This is a bad idea. It'll be abused for maximum powergaming ability by some, and those who need the points to do actual jobs will be at a disadvantage.

Believe it or not, this game isn't just about killing people, so selecting gamer combat stats have a downside of being dead weight if you aren't in combat. Also, I do have plans to add a system to max skills based on jobs, so if it becomes an issue, I can limit combat skills to certain jobs.

**Comment by LazennG** [comment]
image

**Comment by ilikesaying** [comment]
can high cooking skill allow multicrafting of food

**Comment by ilikesaying** [comment]
also 
>Biology: How good you are at researching new designs.

doesn't make sense

**Comment by ghost** [comment]
can someone explain to me why this won’t just widen the skill gap between players, making it easier for valid hunters and antags to murderbone

**Comment by KittyNoodle** [comment]
So far it does not look like antags get any skill boost, meaning that if you don’t have powergame skills normally(with combat skills and medical skills maxed) you are completely out of luck as an antag

**Review by nesisknesis — APPROVED** [review]

**Comment by ghost** [comment]
Skill issue

**Comment by NCRLouTenant** [comment]
Okay, so personally speaking, just because the possibility to minmax is there doesn't mean this shouldn't be added, as there are plenty of people who take perks based on RP over stats (look at the people who use negative quirks)

Plus, if any of it causes too much issues, I'm sure it could be fixed. I'm really into this concept.

**Comment by NCRLouTenant** [comment]
Totally forgot to ask in my previous comment, but any thoughts on possibly giving more perk points for jobs that players have more hours in. Like, tie it to the skill capes hours so that it takes legitimate dedication to a job to earn those extra points?

**Comment by nmajask** [comment]
> Totally forgot to ask in my previous comment, but any thoughts on possibly giving more perk points for jobs that players have more hours in. Like, tie it to the skill capes hours so that it takes legitimate dedication to a job to earn those extra points?

I don't think giving people an advantage for having more hours is a good idea

**Comment by NCRLouTenant** [comment]
> > Totally forgot to ask in my previous comment, but any thoughts on possibly giving more perk points for jobs that players have more hours in. Like, tie it to the skill capes hours so that it takes legitimate dedication to a job to earn those extra points?
> 
> I don't think giving people an advantage for having more hours is a good idea

Fair enough, in a HRP setting I'd argue more for it but here, yeah, I can understand your point.

**Comment by nmajask** [comment]
Closed in favor of #14373
